src/test/scala/is/hail/io/ImportMatrixSuite.scala

```diff
@@ -1,67 +1,171 @@
+import is.hail.annotations.{Annotation, Region, RegionValue, RegionValueBuilder}
 import is.hail.check.Gen
 import is.hail.check.Prop.forAll
 import is.hail.expr._
 import is.hail.expr.types._
+import is.hail.rvd.OrderedRVD
+import is.hail.table.Table
 import is.hail.variant.{MatrixTable, VSMSubgen}
 import org.apache.spark.SparkException
 import org.testng.annotations.Test
 import is.hail.utils._
 
 class ImportMatrixSuite extends SparkSuite {
 
+  val genValidImportType = Gen.oneOf[Type](TInt32(), TInt64(), TFloat32(), TFloat64(), TString())
+
+  val genImportableMatrix = VSMSubgen(
+    sSigGen = Gen.const(TString()),
+    saSigGen = Gen.const(TStruct.empty()),
+    vSigGen = Gen.oneOf[Type](TInt32(), TInt64(), TString()),
+    vaSigGen = Type.preGenStruct(required=false, genValidImportType),
+    globalSigGen = Gen.const(TStruct.empty()),
+    tSigGen = Gen.zip(genValidImportType, Gen.coin(0.2))
+      .map { case (typ, req) => typ.setRequired(req) }.map { t => TStruct("x" -> t) },
+    sGen = (t: Type) => Gen.identifier.map(s => s: Annotation),
+    saGen = (t: Type) => t.genNonmissingValue,
+    vaGen = (t: Type) => t.genNonmissingValue,
+    globalGen = (t: Type) => t.genNonmissingValue,
+    vGen = (t: Type) => t.genNonmissingValue,
+    tGen = (t: Type, v: Annotation) => t.genNonmissingValue)
+
+  def reKeyRows(vsm: MatrixTable): MatrixTable = {
+    val newMatrixType = vsm.matrixType.copy(vType = TInt64())
+    val partStarts = vsm.partitionStarts()
+    val newRowType = newMatrixType.rvRowType
+    val oldRowType = vsm.matrixType.rvRowType
+
+    val indexedRDD = vsm.rdd2.rdd.mapPartitionsWithIndex { case (i, it) =>
+      val region2 = Region()
+      val rv2 = RegionValue(region2)
+      val rv2b = new RegionValueBuilder(region2)
+      var idx = partStarts(i)
+
+      it.map { rv =>
+        region2.clear()
+        rv2b.start(newRowType)
+        rv2b.startStruct()
+        rv2b.addLong(idx)
+        rv2b.addLong(idx)
+        rv2b.addRegionValue(oldRowType.fieldType(2), rv.region, oldRowType.loadField(rv, 2))
+        rv2b.addRegionValue(oldRowType.fieldType(3), rv.region, oldRowType.loadField(rv, 3))
+        rv2b.endStruct()
+        idx += 1
+        rv2.setOffset(rv2b.end())
+        rv2
+      }
+    }
+    new MatrixTable(hc, newMatrixType, vsm.value.localValue, OrderedRVD(newMatrixType.orderedRVType, indexedRDD, None, None))
+  }
+
+  def reKeyCols(vsm: MatrixTable): MatrixTable = {
+    vsm.renameSamples(Array.tabulate[Annotation](vsm.nSamples)("col" + _.toString))
+  }
+
+  def dropRowAnnotations(vsm: MatrixTable, dropFields: Option[Array[String]]=None): MatrixTable =
+    dropFields match {
+      case Some(fields) =>
+        val keep = vsm.vaSignature.fieldNames.filterNot { fn => fields.contains(fn) }
+        val sb = new StringBuilder()
+        keep.foreach { f =>
+          sb ++= s"`$f`: va.`$f`,"
+        }
+        vsm.annotateVariantsExpr(s"va = {${ sb.result().dropRight(1) }}")
```

Comment:
I think when dealing with operations that happen on master / not-per-element we should go with the simplest to read expression, I think here that's:
```scala
val keepExpr = keep.map(f => s"`$f`: va.`$f`").mkString(",")
vsm.annotateVariantsExpr(s"va = { $keepExpr }")
```

I think in particular the dropRight was a bit odd to me at first (because I was still thinking about Arrays)

---

That said this is also a bit of a nitpick so feel free to ignore.